### PR TITLE
[JB-2496] Reload app when files change

### PR DESCRIPTION
**Pull request by stevenwally**
Ticket: [JB-2496](https://juiceanalytics.atlassian.net/browse/JB-2496)
Type: Feature

#### This PR introduces the following changes
- Add `browser-sync` npm package (used to manually refresh a browser instance when file changes are picked up by `jb watch`)
- Add new option to `jb watch` command, `--reload`. This allows users to opt into using an auto reloading browser instance as they develop apps.
- Be selective about when to refresh and when to run `loadjuiceboxapp` (This is actually going away soon in another ticket, [JB-2499](https://juiceanalytics.atlassian.net/browse/JB-2499)).
    - No need to run `loadjuiceboxapp` if file is `*.py`
    - No need to do any refreshing or loading if change is detected in something that is not a file
- Add a small timeout when refreshing after changes in `*.py` files.
    - If Django server hasn't reloaded, just try again until it is ready
    - This may be a little bit more hacky than we want, I'm not sure. Open to suggestions. 
 
## Documentation
When running the `jb watch` command, the option `--reload` may be included. 

`--reload` will instantiate a new browser instance that proxies the Django development server on port 8000. This new browser instance runs on port 3000.

When the watcher detects a change in an application directory, a browser refresh will be triggered in the browser instance that was created when `jb watch --reload` was run.

Some other things to note:
    - If a browser window/tab is pointing to `localhost:8000`, it won't be refreshed as changes are made. This may (or may not) be useful for A/B testing during app development.
    - If multiple browser windows/tabs are pointing to `localhost:3000`, they will all be refreshed in unison when an application file is changed.

## Checklist

- [ ] Add information to the release notes documentation
- [ ] Add new feature to the usage documentation
- [x] Add tests

### CODE REVIEW

File: jbcli/jbcli/utils/reload.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def create_browser_instance():
+    """Create proxy browser instance for hot reloading
+    """
+    cmd = 'npx browser-sync start --proxy="localhost:8000"'
+    os.system(cmd)
+
+
+def refresh_browser(timeout=None):
```
Comment (radix) [inline jbcli/jbcli/utils/reload.py]: I have a nitpick / style suggestion for this function that I think would make the code simpler and shorter.


```python
while True: # or, maybe `for i in range(10)`, if you want to limit how long we'll wait
    if timeout:
        time.sleep(timeout)
    try:
        response = get('http://localhost:8000/health_check')
        if response and response.status_code == 200:
            click.echo('Refreshing browser...')
            cmd = "npx browser-sync reload"
            os.system(cmd)
            return
    except:
        click.echo('Almost done...')
```

This way we don't recurse (which might eventually blow the stack), and we don't need to define the inner `attempt_refresh` function.

File: jbcli/jbcli/utils/reload.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        time.sleep(timeout)
+
+    def attempt_refresh():
+        try:
+            response = get('http://localhost:8000/health_check')
+            if response and response.status_code == 200:
+                click.echo('Refreshing browser...')
+                cmd = "npx browser-sync reload"
+                os.system(cmd)
+        except:
```
Comment (radix) [inline jbcli/jbcli/utils/reload.py]: I would prefer it if we had a more specific exception handler here. What exceptions are you expecting to be caught here? Something from `os.system`, or from the `get` call? I worry that if we get some unexpected error, this code will just loop forever and never print anything to the console indicating what kind of error happened.

So I have two suggestions:

1. only put the `try/except` around the expressions that you expect to raise specific errors
2. be more specific about what exception types you want to catch

File: jbcli/jbcli/utils/reload.py (hunk 1 above)
Comment (stevenwally) [reply jbcli/jbcli/utils/reload.py]: Great suggestion!

File: jbcli/jbcli/utils/reload.py (hunk 2 above)
Comment (stevenwally) [reply jbcli/jbcli/utils/reload.py]: Ah, that absolutely makes sense. Will address.

File: jbcli/jbcli/utils/reload.py [hunk 3]
```diff
@@ -22,20 +22,18 @@ def refresh_browser(timeout=None):
...
     click.echo('Checking server status...')
-
-    if timeout:
-        time.sleep(timeout)
-
-    def attempt_refresh():
+    for i in range(5):
+        if timeout:
+            time.sleep(timeout)
+        response = {'status_code': None}
```
Comment (radix) [inline jbcli/jbcli/utils/reload.py]: this doesn't seem to be a meaningful line, since it will immediately get overwritten with the result of `get`.

Comment (stevenwally) [reply jbcli/jbcli/utils/reload.py]: Whoops! Meant to remove this. Thanks.

**Pull request merged**